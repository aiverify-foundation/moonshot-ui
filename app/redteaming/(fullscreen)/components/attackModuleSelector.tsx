import { IconName } from '@/app/components/IconSVG';
import { useGetAllAttackModulesQuery } from '@/app/services/attack-modules-api-service';
import { SelectListItem } from '@/app/views/shared-components/selectListItem';
import tailwindConfig from '@/tailwind.config';
const colors = tailwindConfig.theme?.extend?.colors as CustomColors;

type AttackModuleSelectViewProps = {
  selectedAttack?: AttackModule;
  onAttackClick: (attack: AttackModule) => void;
  onSkipClick?: () => void;
};

function AttackModuleSelectView(props: AttackModuleSelectViewProps) {
  const { selectedAttack, onAttackClick } = props;
  const { data, isFetching } = useGetAllAttackModulesQuery();

  return (
    <div className="flex flex-col pt-4 gap-8 pb-4 h-[80%] items-center">
      <section className="flex flex-col items-center gap-3">
        <hgroup>
          <h2 className="text-[1.6rem] leading-[2rem] height-[2rem] font-medium tracking-wide text-white w-full text-center">
            Would you like to use any of these attack modules?
          </h2>
          <p className="text-[0.9rem] text-moongray-400 text-center">
            <span className="font-bold">Optional</span>&nbsp;modules that
            automatically generate prompts to red team an Endpoint.
          </p>
        </hgroup>
      </section>
      <div className="relative flex flex-col min-h-[300px] px-[10%] w-[100%] h-full items-center">
        {isFetching || !data ? (
          <div className="ring">
            Loading
            <span />
          </div>
        ) : (
          <>
            <ul
              className="flex flex-row flex-wrap gap-[2%] w-[100%] overflow-y-auto custom-scrollbar px-4"
              style={{ height: '100%' }}>
              {data.map((attack) => {
                const isSelected =
                  selectedAttack && selectedAttack.id === attack.id;
                const description = attack.description
                  ? attack.description
                      .split('Parameters:')[0]
                      .replace(/\n/g, '<br />')
                  : '';
                return (
                  <SelectListItem<AttackModule>
                    key={attack.id}
                    label={attack.name}
                    onClick={() => onAttackClick(attack)}
                    iconName={IconName.OutlineBox}
                    item={attack}
                    hideCheckbox
                    bgColor={isSelected ? colors.moonwine[950] : undefined}
                    style={{
                      alignItems: 'flex-start',
                      height: 180,
                      overflowY: 'hidden',
                    }}>
                    <p
                      className="text-[0.8rem] text-moongray-400"
                      dangerouslySetInnerHTML={{ __html: description }}
                    />
                  </SelectListItem>
                );
              })}
            </ul>
          </>
        )}
      </div>
    </div>
  );
}

export { AttackModuleSelectView };
