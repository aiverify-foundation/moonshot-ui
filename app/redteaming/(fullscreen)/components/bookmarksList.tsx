function BookmarksList() {
  return (
    <div className="flex flex-col min-h-[300px]">
      <h3 className="mb-5 text-white">Bookmarks</h3>
      <ul className="list-none bg-white rounded-sm">
        <li className="p-1 border-b">
          <p>Bookmark 1</p>
        </li>
        <li className="p-1 border-b">
          <p>Bookmark 2</p>
        </li>
        <li className="p-1 border-b">
          <p>Bookmark 3</p>
        </li>
        <li className="p-1">
          <p>Bookmark 4</p>
        </li>
      </ul>
    </div>
  );
}

export { BookmarksList };
