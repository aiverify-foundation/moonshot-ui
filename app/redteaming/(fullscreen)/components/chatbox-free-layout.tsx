import React, { useEffect, useState } from 'react';
import { colors } from '@/app/customColors';
import {
  getWindowId,
  getWindowScrollTopById,
  getWindowSizeById,
  getWindowXYById,
} from '@/app/lib/window-utils';
import useChatboxesPositionsUtils from '@/app/redteaming/(fullscreen)/hooks/useChatboxesPositionsUtils';
import { useAppSelector } from '@/lib/redux';
import { ChatBox, ChatBoxControls } from './chatbox';

const minimizedStyle = {
  transform: 'scale(0.1)',
  width: 500,
  height: 500,
  top: `calc(100vh - 350px)`,
  transition:
    'transform 0.3s ease-in, top 0.4s ease-in-out, left 0.5s ease-out', // Smooth transition for the animation
};

type ChatFreeLayoutProps = {
  chatSession: SessionData;
  chatCompletionInProgress: boolean;
  isAttackMode: boolean;
  promptTemplates: PromptTemplate[];
  selectedPromptTemplate: PromptTemplate | undefined;
  promptText: string;
  handleOnWindowChange: (
    x: number,
    y: number,
    width: number,
    height: number,
    scrollPosition: number,
    windowId: string
  ) => void;
  handleCreatePromptBookmarkClick: CreatePromptBookmarkFunction;
};

const ChatboxFreeLayout = React.forwardRef(
  (
    props: ChatFreeLayoutProps,
    ref: React.Ref<Map<string, ChatBoxControls>>
  ) => {
    const {
      chatSession,
      chatCompletionInProgress,
      isAttackMode,
      promptTemplates,
      selectedPromptTemplate,
      promptText,
      handleOnWindowChange,
      handleCreatePromptBookmarkClick,
    } = props;
    const windowsMap = useAppSelector((state) => state.windows.map);
    const { resetChatboxPositions } = useChatboxesPositionsUtils(chatSession);
    const [minizedChats, setMinizedChats] = useState<string[]>([]);
    const [isMaximizing, setIsMazimizing] = useState(false);
    const chatBoxControlsMap = new Map<string, ChatBoxControls>();
    React.useImperativeHandle(ref, () => chatBoxControlsMap);

    function handleMinimizeClick(windowId: string) {
      return () => {
        setMinizedChats([...minizedChats, windowId]);
      };
    }

    function handleMaximizeClick(windowId: string) {
      return () => {
        setIsMazimizing(true);
        setMinizedChats(minizedChats.filter((id) => id !== windowId));
      };
    }

    useEffect(() => {
      resetChatboxPositions(false);
    }, []);

    useEffect(() => {
      setTimeout(() => {
        setIsMazimizing(false);
      }, 300);
    }, [isMaximizing]);

    const chatboxHeaderStyle: React.CSSProperties = {
      borderTopLeftRadius: '0.5rem',
      borderTopRightRadius: '0.5rem',
    };

    return (
      <>
        {chatSession.session.endpoints.map((id: string, index: number) => {
          const isMinimized = minizedChats.includes(getWindowId(id));
          const left = index * 20;
          const chatboxStyle = isMinimized
            ? { ...minimizedStyle, left }
            : isMaximizing
              ? {
                  transition:
                    'transform 0.3s ease-in-out, top 0.5s ease-in-out, left 0.5s ease-in-out',
                }
              : { borderRadius: '0.5rem' };
          return windowsMap[getWindowId(id)] ? (
            <ChatBox
              key={id}
              disableCloseIcon={
                chatSession.session.endpoints.length === 1 ||
                chatSession.session.endpoints.length - minizedChats.length === 1
              }
              resizable
              draggable
              disableOnScroll
              disableBubbleTooltips={isMinimized}
              resizeHandlerColor={colors.moonwine[500]}
              ref={(imperativeHandles) => {
                if (imperativeHandles) {
                  chatBoxControlsMap.set(id, imperativeHandles);
                } else {
                  chatBoxControlsMap.delete(id);
                }
              }}
              windowId={getWindowId(id)}
              title={id}
              initialXY={getWindowXYById(windowsMap, id)}
              initialSize={getWindowSizeById(windowsMap, id)}
              initialScrollTop={getWindowScrollTopById(windowsMap, id)}
              chatHistory={
                chatSession.chat_records
                  ? chatSession.chat_records[id] || []
                  : []
              }
              promptTemplates={promptTemplates}
              currentPromptTemplate={selectedPromptTemplate}
              isAttackMode={isAttackMode}
              currentPromptText={promptText}
              isChatCompletionInProgress={chatCompletionInProgress}
              onWindowChange={handleOnWindowChange}
              styles={chatboxStyle}
              headerStyle={chatboxHeaderStyle}
              onCloseClick={handleMinimizeClick(getWindowId(id))}
              onWholeWindowClick={
                isMinimized ? handleMaximizeClick(getWindowId(id)) : undefined
              }
              onCreatePromptBookmarkClick={handleCreatePromptBookmarkClick}
            />
          ) : null;
        })}
      </>
    );
  }
);

ChatboxFreeLayout.displayName = 'ChatboxFreeLayout';

export { ChatboxFreeLayout };
