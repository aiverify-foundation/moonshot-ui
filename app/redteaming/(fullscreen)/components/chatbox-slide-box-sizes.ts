export type SlideChatBoxDimensions = {
  width: number;
  height: number;
  gap: number;
};

export function getDefaultChatBoxSizes() {
  // IMPORTANT: Also align the values of --chatwindow-width and --gap-width css variables in global.css

  // Default values
  let width = 420;
  let height = 550;
  let gap = 16;

  if (typeof window === 'undefined') {
    return { width: 0, height: 0, gap: 0 };
  }

  // Media query for large screens (e.g., more than 1180px)
  if (window.matchMedia('(min-width: 1180px)').matches) {
    console.log('1180px');
    width = 420; // Smaller width for small screens
    height = 500; // Smaller height for small screens
    gap = 50; // Smaller gap for small screens
  }

  // Media query for medium screens (e.g., less than 992px)
  if (window.matchMedia('(max-width: 992px)').matches) {
    console.log('992px');
    width = 420; // Smaller width for small screens
    height = 450; // Smaller height for small screens
    gap = 15; // Smaller gap for small screens
  }

  // Media query for small screens (e.g., less than 768px)
  if (window.matchMedia('(max-width: 768px)').matches) {
    console.log('768px');
    width = 300; // Smaller width for small screens
    height = 350; // Smaller height for small screens
    gap = 15; // Smaller gap for small screens
  }

  // Media query for very small screens (e.g., less than 480px)
  if (window.matchMedia('(max-width: 480px)').matches) {
    console.log('480px');
    width = 280; // Even smaller width for very small screens
    height = 300; // Even smaller height for very small screens
    gap = 10; // Even smaller gap for very small screens
  }

  // Return updated values
  return { width, height, gap };
}
