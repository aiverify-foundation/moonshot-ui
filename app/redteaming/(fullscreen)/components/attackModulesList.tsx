import React from 'react';
import { Icon, IconName } from '@/app/components/IconSVG';
import { Button, ButtonType } from '@/app/components/button';
import { useGetAllAttackModulesQuery } from '@/app/services/attack-modules-api-service';
import { colors } from '@/app/views/shared-components/customColors';
import { LoadingAnimation } from '@/app/views/shared-components/loadingAnimation';

type AttackModulesListProps = {
  onPrimaryBtnClick: (attackModule: AttackModule) => void;
  onSecondaryBtnClick?: () => void;
};

function AttackModulesList(props: AttackModulesListProps) {
  const { onPrimaryBtnClick, onSecondaryBtnClick } = props;
  const [selectedAttackModule, setSelectedAttackModule] = React.useState<
    AttackModule | undefined
  >();
  const { data, refetch, isLoading } = useGetAllAttackModulesQuery();

  React.useEffect(() => {
    refetch();
  }, []);

  let attackModuleDetailsSection: React.ReactNode;

  if (selectedAttackModule) {
    const description = selectedAttackModule.description
      ? selectedAttackModule.description.replace(/\n/g, '<br /><br/>')
      : '';

    attackModuleDetailsSection = (
      <section className="p-2 bg-moongray-950 w-[480px]">
        <header className="flex flex-col gap-2 mb-4">
          <div className="flex gap-2">
            <Icon name={IconName.MoonAttackStrategy} />
            <h3 className="text-[1rem] font-semibold text-white">
              {selectedAttackModule.name}
            </h3>
          </div>
        </header>
        <p
          dangerouslySetInnerHTML={{ __html: description }}
          className="text-[0.85rem] overflow-y-auto custom-scrollbar h-full text-moongray-200"
          style={{ height: 'calc(100% - 40px)' }}
        />
      </section>
    );
  }

  const listOfModules = data ? (
    <ul className="divide-y divide-moongray-500 max-w-[400px] overflow-x-hidden overflow-y-auto custom-scrollbar h-full">
      {data.map((attackModule) => {
        const isSelected =
          selectedAttackModule && selectedAttackModule.id === attackModule.id;
        const description = attackModule.description
          ? attackModule.description
              .split('Parameters:')[0]
              .replace(/\n/g, '<br />')
          : '';
        return (
          <li
            key={attackModule.id}
            className={`p-2 ${!isSelected && 'hover:bg-moongray-900'} active:bg-moongray-600 cursor-pointer 
            ${isSelected ? 'bg-moongray-950' : 'bg-moongray-700'}`}
            onClick={() => setSelectedAttackModule(attackModule)}
            style={{
              transition: 'background-color 0.2s ease-in-out',
            }}>
            <header className="flex flex-col gap-2">
              <div className="flex gap-2">
                <Icon name={IconName.MoonAttackStrategy} />
                <h3 className="text-[0.87rem] text-white">
                  {attackModule.name}
                </h3>
              </div>
              <p
                className="text-sm h-[80px] min-h-[80px] overflow-y-hidden text-[0.74rem] text-moongray-200"
                dangerouslySetInnerHTML={{ __html: description }}
              />
            </header>
          </li>
        );
      })}
    </ul>
  ) : null;

  return (
    <div className="relative flex flex-col w-full h-full gap-4">
      {isLoading || !data ? (
        <LoadingAnimation />
      ) : (
        <>
          <main
            className="h-full"
            style={{ height: 'calc(100% - 70px)' }}>
            <div className="flex gap-4 h-full">
              {listOfModules}
              {attackModuleDetailsSection}
            </div>
          </main>
          <footer className="flex justify-end gap-2">
            <Button
              mode={ButtonType.OUTLINE}
              onClick={onSecondaryBtnClick}
              text="Cancel"
              hoverBtnColor={colors.moongray[700]}
              pressedBtnColor={colors.moongray[800]}
            />
            <Button
              disabled={selectedAttackModule === undefined}
              mode={ButtonType.PRIMARY}
              onClick={() =>
                selectedAttackModule
                  ? onPrimaryBtnClick(selectedAttackModule)
                  : null
              }
              text="Use"
              hoverBtnColor={colors.moongray[950]}
            />
          </footer>
        </>
      )}
    </div>
  );
}

export { AttackModulesList };
