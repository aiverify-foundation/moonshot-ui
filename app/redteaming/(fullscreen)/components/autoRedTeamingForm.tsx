import { ChangeEvent, useState } from 'react';
import { Button, ButtonType } from '@/app/components/button';
import { SelectInput, SelectOption } from '@/app/components/selectInput';
import { TextArea } from '@/app/components/textArea';
import { useGetAllRecipesQuery } from '@/app/services/recipe-api-service';
import tailwindConfig from '@/tailwind.config';

const colors = tailwindConfig.theme?.extend?.colors as CustomColors;

function AutoRedTeamingForm() {
  const [selectedRecipe, setSelectedRecipe] = useState<Recipe | undefined>(
    undefined
  );
  const [initialPrompt, setInitialPrompt] = useState<string>('');

  const { data: recipes, isLoading: recipesIsLoading } =
    useGetAllRecipesQuery(undefined);

  let recipesSelectionOptions: SelectOption[] = [];
  if (recipes && recipes.length) {
    recipesSelectionOptions = recipes.map((rcp) => ({
      value: rcp.id,
      label: rcp.name,
    }));
  }

  function handleChange(
    e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) {
    if (e.target.name === 'recipe') {
      setSelectedRecipe(recipes?.find((rcp) => rcp.id === e.target.value));
    } else if (e.target.name === 'initialPrompt') {
      setInitialPrompt(e.target.value);
    }
  }

  return (
    <div className="relative flex flex-col min-h-[300px]">
      <h3 className="mb-5 text-white">Automated Red Teaming</h3>
      {recipesIsLoading ? (
        <div className="ring">
          Loading
          <span />
        </div>
      ) : (
        <>
          <SelectInput
            label="Select Recipe"
            name="recipe"
            value={selectedRecipe?.id}
            placeholder="Select a recipe"
            options={recipesSelectionOptions}
            onSyntheticChange={handleChange}
          />
          <TextArea
            name="initialPrompt"
            label="Initial Prompt"
            onChange={handleChange}
            value={initialPrompt}
            placeholder="Enter initial prompt here"
          />
          <div className="flex justify-left mt-2">
            <Button
              type="submit"
              mode={ButtonType.PRIMARY}
              text="Start Red Team"
              btnColor={colors.moongray[950]}
              hoverBtnColor={colors.moongray[900]}
              textColor={colors.white}
            />
          </div>
        </>
      )}
    </div>
  );
}

export { AutoRedTeamingForm };
