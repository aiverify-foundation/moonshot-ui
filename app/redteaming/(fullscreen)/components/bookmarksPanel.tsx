import React from 'react';
import { IconName } from '@/app/components/IconSVG';
import { Button, ButtonType } from '@/app/components/button';
import { colors } from '@/app/customColors';
import { ViewBookmarksModal } from './viewBookmarksModal';

function BookmarksPanel({
  bottom,
  left,
  disabled,
  defaultShowPanel = false,
  onUseBtnClick,
  onPanelClose,
}: {
  bottom: React.CSSProperties['top'];
  left: React.CSSProperties['left'];
  disabled: boolean;
  defaultShowPanel?: boolean;
  onPanelClose?: () => void;
  onUseBtnClick: (preparedPrompt: string) => void;
}) {
  const [showPanel, setShowPanel] = React.useState(() => defaultShowPanel);

  React.useEffect(() => {
    setShowPanel(defaultShowPanel);
  }, [defaultShowPanel]);

  function handleUseBtnClick(preparedPrompt: string) {
    onUseBtnClick(preparedPrompt);
    setShowPanel(false);
    if (onPanelClose) {
      onPanelClose();
    }
  }

  function handleCancelBtnClick() {
    setShowPanel(false);
    if (onPanelClose) {
      onPanelClose();
    }
  }

  return (
    <div
      className="bg-moongray-600 w-[220px] absolute rounded-md p-2 shadow-lg
        flex gap-4"
      style={{ bottom, left }}>
      {disabled && (
        <div
          className="absolute gap-2 bg-moongray-950/50 w-full h-full z-10
            flex justify-center items-center rounded-md"
          style={{ top: 0, left: 0 }}>
          <div className="waitspinner" />
        </div>
      )}
      {showPanel && (
        <ViewBookmarksModal
          onCloseIconClick={handleCancelBtnClick}
          onPrimaryBtnClick={handleUseBtnClick}
        />
      )}
      <div className="flex items-center gap-2">
        <Button
          width={200}
          alignContent="flex-start"
          text="Bookmarked Prompts"
          type="button"
          mode={ButtonType.TEXT}
          hoverBtnColor={colors.moongray[500]}
          pressedBtnColor={colors.moongray[400]}
          leftIconName={IconName.Ribbon}
          onClick={() => setShowPanel(true)}
        />
      </div>
    </div>
  );
}

export { BookmarksPanel };
