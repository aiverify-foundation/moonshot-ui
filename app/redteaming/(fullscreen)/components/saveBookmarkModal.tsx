import React from 'react';
import { useFormState } from 'react-dom';
import { createBookmark } from '@/actions/createBookmark';
import { Icon, IconName } from '@/app/components/IconSVG';
import { Button, ButtonType } from '@/app/components/button';
import { Chat } from '@/app/components/chat';
import { Modal } from '@/app/components/modal';
import { TextInput } from '@/app/components/textInput';
import { colors } from '@/app/customColors';
import { PromptBubbleInfo } from './prompt-bubble-info';
import { SubmitButton } from './submitButton';

type SaveBookMarkModalProps = {
  duration: string;
  prompt: string;
  preparedPrompt: string;
  attackModule: string | undefined;
  contextStrategy: string | undefined;
  promptTemplateName: string | undefined;
  template: string | undefined;
  response: string;
  metric: string | undefined;
  onCloseIconClick: () => void;
  onPrimaryBtnClick: () => void;
};

const initialFormValues: FormState<BookmarkFormValues> = {
  formStatus: 'initial',
  formErrors: undefined,
  name: '',
  prompt_template: '',
  prepared_prompt: '',
  response: '',
  metric: undefined,
  attack_module: undefined,
  context_strategy: undefined,
};

function SaveBookMarkModal(props: SaveBookMarkModalProps) {
  const {
    duration,
    prompt,
    preparedPrompt,
    promptTemplateName,
    template,
    response,
    metric,
    attackModule,
    contextStrategy,
    onCloseIconClick,
    onPrimaryBtnClick,
  } = props;
  const [bookmarkName, setBookmarkName] = React.useState('');
  const [showResultModal, setShowResultModal] = React.useState(false);
  const [showErrorModal, setShowErrorModal] = React.useState(false);
  const [formState, action] = useFormState<
    FormState<BookmarkFormValues>,
    FormData
  >(createBookmark, initialFormValues);

  React.useEffect(() => {
    if (formState.formStatus === 'error') {
      setShowErrorModal(true);
      return;
    }
    if (formState.formStatus === 'success') {
      setShowResultModal(true);
      setBookmarkName('');
    }
  }, [formState]);

  function handleBookmarkNameChange(e: React.ChangeEvent<HTMLInputElement>) {
    setBookmarkName(e.target.value);
  }

  return (
    <>
      {showResultModal ? (
        <Modal
          heading="Bookmark Saved"
          bgColor={colors.moongray['800']}
          textColor="#FFFFFF"
          primaryBtnLabel="View Bookmarks"
          enableScreenOverlay
          overlayOpacity={0.8}
          onCloseIconClick={onCloseIconClick}
          onPrimaryBtnClick={onPrimaryBtnClick}>
          <div className="flex gap-2 items-start">
            <p>{`Bookmark ${formState.name} was successfully saved.`}</p>
          </div>
        </Modal>
      ) : null}
      {showErrorModal ? (
        <Modal
          heading="Errors"
          bgColor={colors.moongray['800']}
          textColor="#FFFFFF"
          primaryBtnLabel="Close"
          enableScreenOverlay
          overlayOpacity={0.8}
          onCloseIconClick={() => setShowErrorModal(false)}
          onPrimaryBtnClick={() => setShowErrorModal(false)}>
          <div className="flex gap-2 items-start">
            <Icon
              name={IconName.Alert}
              size={40}
              color="red"
            />
            {formState.formErrors ? (
              <ul>
                {Object.entries(formState.formErrors).map(([key, value]) => (
                  <li key={key}>
                    {key}: {value.join(', ')}
                  </li>
                ))}
              </ul>
            ) : (
              'An unknown error occurred'
            )}
          </div>
        </Modal>
      ) : null}
      {showResultModal || showErrorModal ? null : (
        <Modal
          width={900}
          height={640}
          heading="Save New Bookmark"
          bgColor={colors.moongray['800']}
          textColor="#FFFFFF"
          enableScreenOverlay={true}
          overlayOpacity={0.8}
          onCloseIconClick={onCloseIconClick}>
          <p className="mb-8 text-[0.95rem]">
            Save this prompt as a bookmark to be able to quickly reference it in
            future red teaming exercises.
          </p>
          <div className="flex gap-4 h-[300px]">
            <section className="flex-1 h-full bg-chatboxbg p-4 pr-2 rounded-lg">
              <div className="custom-scrollbar overflow-y-auto h-full">
                <div className="max-w-[90%]">
                  <h2 className="text-left text-white text-[0.8rem] mb-2">
                    Prompt
                  </h2>
                  <Chat.TalkBubble
                    backgroundColor={colors.imdapurple}
                    textAlign="left"
                    fontColor="#FFF"
                    fontSize={14}
                    marginBottom={0}>
                    {preparedPrompt}
                  </Chat.TalkBubble>
                  <h2 className="text-left text-white text-[0.8rem] mb-2 mt-4">
                    Response
                  </h2>
                  <Chat.TalkBubble
                    backgroundColor={colors.chatBubbleWhite}
                    textAlign="left"
                    fontColor={colors.black}
                    fontSize={14}
                    marginBottom={0}>
                    {response}
                  </Chat.TalkBubble>
                </div>
              </div>
            </section>
            <section className="flex-1">
              <PromptBubbleInfo
                duration={duration}
                prompt={preparedPrompt}
                prompt_template={promptTemplateName}
                templateString={template}
              />
            </section>
          </div>
          <form
            action={action}
            className="absolute bottom-0 left-0 flex
          flex-col gap-2 p-7 w-full mt-4">
            <div className="border-t border-moongray-500 w-[98%] mb-2" />
            <div className="w-[380px]">
              <TextInput
                id="bookmarkName"
                name="name"
                label="Name *"
                labelStyles={{
                  fontSize: '0.9rem',
                  color: colors.moonpurplelight,
                }}
                inputStyles={{ height: 38 }}
                placeholder="Give this bookmark a unique name"
                value={bookmarkName}
                onChange={handleBookmarkNameChange}
                error={formState.formErrors?.name?.[0]}
              />
            </div>
            <input
              type="hidden"
              name="prompt"
              value={prompt}
            />
            <input
              type="hidden"
              name="prompt_template"
              value={template}
            />
            <input
              type="hidden"
              name="prepared_prompt"
              value={preparedPrompt}
            />
            <input
              type="hidden"
              name="response"
              value={response}
            />
            <input
              type="hidden"
              name="metric"
              value={metric}
            />
            <input
              type="hidden"
              name="attack_module"
              value={attackModule}
            />
            <input
              type="hidden"
              name="context_strategy"
              value={contextStrategy}
            />
            <div className="flex gap-4 justify-end">
              <Button
                mode={ButtonType.OUTLINE}
                size="lg"
                width={120}
                text="Cancel"
                hoverBtnColor={colors.moongray[700]}
                pressedBtnColor={colors.moongray[800]}
                onClick={onCloseIconClick}
              />
              <SubmitButton disabled={bookmarkName === ''} />
            </div>
          </form>
        </Modal>
      )}
    </>
  );
}

export { SaveBookMarkModal };
