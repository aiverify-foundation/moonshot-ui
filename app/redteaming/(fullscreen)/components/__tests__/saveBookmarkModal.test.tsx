import { SaveBookMarkModal } from '@/app/redteaming/(fullscreen)/components/saveBookmarkModal';
import { render, screen } from '@testing-library/react';

describe('SaveBookmarkModal', () => {
  it('should render', () => {
    const mockProps = {
      duration: '2s',
      prompt: 'Test prompt',
      preparedPrompt: 'Prepared test prompt',
      attackModule: 'Test attack module',
      contextStrategy: 'Test context strategy',
      promptTemplateName: 'Test template',
      template: 'Test template string',
      response: 'Test response',
      metric: 'Test metric',
      onCloseIconClick: jest.fn(),
    };

    const { container } = render(<SaveBookMarkModal {...mockProps} />);
  });
});
