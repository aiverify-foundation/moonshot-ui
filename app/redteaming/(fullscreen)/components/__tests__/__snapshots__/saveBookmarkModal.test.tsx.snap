// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SaveBookmarkModal renders and populates hidden form elements with the correct values 1`] = `
<div>
  <div
    class="fixed inset-0 flex items-center justify-center bg-black z-[999]"
    style="opacity: 0.8;"
  />
  <div
    class="absolute top-1/2 left-1/2 transform -translate-x-1/2 -translate-y-1/2 w-[600px] h-[260px] z-[1000] rounded-[15px] p-6"
    style="width: 900px; height: 640px; background-color: rgb(70, 67, 73); color: rgb(255, 255, 255); box-shadow: 0 0 10px rgba(0,0,0,.5);"
  >
    <header
      class="flex justify-between mb-3"
    >
      <h1
        class="text-[1.4rem] font-normal"
      >
        Save New Bookmark
      </h1>
      <div
        class="
        flex items-center justify-center 
        cursor-pointer
        hover:opacity-50 active:opacity-25
        
      "
      >
        <svg
          height="20"
          viewBox="0 0 32 32"
          width="20"
        >
          <g
            id="cross"
          >
            <line
              stroke="#FFFFFF"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              x1="7"
              x2="25"
              y1="7"
              y2="25"
            />
            <line
              stroke="#FFFFFF"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              x1="7"
              x2="25"
              y1="25"
              y2="7"
            />
          </g>
        </svg>
      </div>
    </header>
    <main>
      <p
        class="mb-8 text-[0.95rem]"
      >
        Save this turn as a bookmark to be able to quickly reference it in future red teaming exercises.
      </p>
      <div
        class="flex gap-4 h-[300px]"
      >
        <section
          class="flex-1 h-full bg-chatboxbg p-4 pr-2 rounded-lg"
        >
          <div
            class="custom-scrollbar overflow-y-auto h-full"
          >
            <div
              class="max-w-[90%]"
            >
              <h2
                class="text-left text-white text-[0.8rem] mb-2"
              >
                Prompt
              </h2>
              <div
                style="text-align: left; color: rgb(255, 255, 255); padding: 12px 16px; font-size: 14px; background-color: rgb(73, 48, 90); margin: 0px 0px 0px 0px; border-radius: 14px; min-width: 35%;"
              >
                Prepared test prompt
              </div>
              <h2
                class="text-left text-white text-[0.8rem] mb-2 mt-4"
              >
                Response
              </h2>
              <div
                style="text-align: left; color: rgb(0, 0, 0); padding: 12px 16px; font-size: 14px; background-color: rgb(221, 221, 221); margin: 0px 0px 0px 0px; border-radius: 14px; min-width: 35%;"
              >
                Test response
              </div>
            </div>
          </div>
        </section>
        <section
          class="flex-1"
        >
          <div
            class="w-full p-2"
          >
            <h3
              class="mb-1 underline text-stone-50"
            >
              Prompt Details
            </h3>
            <div>
              <span
                class="text-sky-400 pr-1"
              >
                Response Time Taken (secs):
              </span>
              <span
                class="text-white"
              >
                2.0000
              </span>
            </div>
            <div>
              <span
                class="text-sky-400 pr-1"
              >
                Template Name:
              </span>
              <span
                class="text-white"
              >
                Test template
              </span>
            </div>
            <div
              class="text-stone-50 dark:text-sky-400 mb-1 mt-2 underline"
            >
              Template String
            </div>
            <div
              class="max-h-[200px] custom-scrollbar overflow-y-auto"
            >
              <div>
                <span
                  style="color: rgb(207, 250, 254);"
                >
                  Test template string
                </span>
              </div>
            </div>
          </div>
        </section>
      </div>
      <form
        action="unused"
        class="absolute bottom-0 left-0 flex flex-col gap-2 p-7 w-full mt-4"
      >
        <div
          class="border-t border-moongray-500 w-[98%] mb-2"
        />
        <div
          class="w-[380px]"
        >
          <div
            class="textInput"
          >
            <label
              for="bookmarkName"
            >
              <div
                class="label"
                style="font-size: 0.9rem; color: rgb(213, 170, 234);"
              >
                <div>
                  Name *
                </div>
              </div>
              <input
                autocomplete="off"
                id="bookmarkName"
                name="name"
                placeholder="Give this bookmark a unique name"
                style="height: 38px;"
                type="text"
                value=""
              />
            </label>
          </div>
        </div>
        <input
          name="prompt"
          type="hidden"
          value="Test prompt"
        />
        <input
          name="prompt_template"
          type="hidden"
          value="Test template string"
        />
        <input
          name="prepared_prompt"
          type="hidden"
          value="Prepared test prompt"
        />
        <input
          name="response"
          type="hidden"
          value="Test response"
        />
        <input
          name="metric"
          type="hidden"
          value="Test metric"
        />
        <input
          name="attack_module"
          type="hidden"
          value="Test attack module"
        />
        <input
          name="context_strategy"
          type="hidden"
          value="Test context strategy"
        />
        <div
          class="flex gap-4 justify-end"
        >
          <button
            class="btn btn_outline btn_lg"
            style="width: 120px;"
            type="button"
          >
            <span>
              Cancel
            </span>
          </button>
          <button
            class="btn btn_primary btn_lg"
            disabled=""
            style="width: 120px;"
            type="submit"
          >
            <span>
              Save
            </span>
          </button>
        </div>
      </form>
    </main>
  </div>
</div>
`;
